File: app/src/main/java/com/example/habitpalette/ui/home/CurrentHabitRecyclerAdapter.java
package com.example.habitpalette.ui.home;

import android.content.Context;
import android.view.LayoutInflater;
import android.view.View;
import android.view.ViewGroup;
import android.widget.TextView;

import androidx.annotation.NonNull;
import androidx.recyclerview.widget.RecyclerView;

import com.example.habitpalette.ui.common.HabitScoreProgressView;
import com.example.habitpalette.R;

import java.text.ParseException;
import java.text.SimpleDateFormat;
import java.util.ArrayList;
import java.util.Date;
import java.util.Locale;

public class CurrentHabitRecyclerAdapter extends RecyclerView.Adapter<CurrentHabitRecyclerAdapter.ViewHolder> {

    private ArrayList<CurrentHabitRecyclerItem> mData;

    CurrentHabitRecyclerAdapter(ArrayList<CurrentHabitRecyclerItem> list) {
        mData = list;
    }

    @NonNull
    @Override
    public ViewHolder onCreateViewHolder(@NonNull ViewGroup parent, int viewType) {
        Context context = parent.getContext();
        LayoutInflater inflater = (LayoutInflater) context.getSystemService(Context.LAYOUT_INFLATER_SERVICE);

        View view = inflater.inflate(R.layout.item_current_habit, parent, false);
        CurrentHabitRecyclerAdapter.ViewHolder vh = new CurrentHabitRecyclerAdapter.ViewHolder(view);

        return vh;
    }

    @Override
    public void onBindViewHolder(@NonNull ViewHolder holder, int position) {
        CurrentHabitRecyclerItem item = mData.get(position);
        SimpleDateFormat sdf = new SimpleDateFormat("yyyy.MM.dd", Locale.KOREA);
        Date today = new Date();
        int compare=0;
        try {
            Date currentDate = sdf.parse(sdf.format(today));
            Date startDate = sdf.parse(sdf.format(item.getHabitPeriod()));
            compare = currentDate.compareTo(startDate);
             } catch (ParseException e) {
            e.printStackTrace();
        }
        int score = (int) item.getHabitScore()*20;
        holder.habitTitle.setText(item.getHabitTitle());
        holder.habitScore.setPercent(score);
        holder.habitScore.setText(Float.toString(item.getHabitScore()));
        holder.habitScore.setFgColorStart(item.getHabitColor());
        holder.habitScore.setFgColorEnd(0xff65b8b5);
        holder.habitPeriod.setText("습관 "+Integer.toString(compare)+"일째");
    }

    @Override
    public int getItemCount() {
        return mData.size();
    }

    public class ViewHolder extends RecyclerView.ViewHolder {
        TextView habitTitle;
        TextView habitPeriod;
        HabitScoreProgressView habitScore;

        ViewHolder(View itemView) {
            super(itemView);

            habitTitle = itemView.findViewById(R.id.text_habit_title);
            habitPeriod = itemView.findViewById(R.id.text_habit_period);
            habitScore = itemView.findViewById(R.id.progress_view_current_habit_score);
        }
    }
}

File: app/src/main/java/com/example/habitpalette/ui/home/CurrentHabitRecyclerAdapter.kt
package com.example.habitpalette.ui.home

import android.annotation.SuppressLint
import android.os.Build
import androidx.recyclerview.widget.RecyclerView
import android.view.ViewGroup
import android.view.LayoutInflater
import androidx.annotation.RequiresApi
import androidx.recyclerview.widget.DiffUtil
import androidx.recyclerview.widget.ListAdapter
import com.example.habitpalette.data.model.Habit
import com.example.habitpalette.databinding.ItemCurrentHabitBinding
import java.time.LocalDate
import java.time.temporal.ChronoUnit

class CurrentHabitRecyclerAdapter:
    ListAdapter<Habit, CurrentHabitRecyclerAdapter.CurrentHabitViewHolder>(CurrentHabitComparator()){

    private lateinit var itemBinding: ItemCurrentHabitBinding

    inner class CurrentHabitViewHolder(private val itemBinding: ItemCurrentHabitBinding): RecyclerView.ViewHolder(itemBinding.root){
        private var habit:Habit? = null

        // item 클릭해서 CalendarActivity로 intent

        @SuppressLint("SetTextI18n")
        @RequiresApi(Build.VERSION_CODES.O)
        fun bind(data: Habit) {
            itemBinding.tvTitle.text = data.title
            itemBinding.tvPeriod.text = ChronoUnit.DAYS.between(data.start_date, LocalDate.now()).toString()+"일째"
            itemBinding.pvCurrentProgress.percent = 50f
        }
    }

    override fun onCreateViewHolder(parent: ViewGroup, viewType: Int): CurrentHabitViewHolder {
        itemBinding = ItemCurrentHabitBinding.inflate(LayoutInflater.from(parent.context), parent, false)
        return CurrentHabitViewHolder(itemBinding)
    }

    @RequiresApi(Build.VERSION_CODES.O)
    override fun onBindViewHolder(holder: CurrentHabitViewHolder, position: Int) {
        holder.bind(getItem(position))
    }

    // Room을 사용할 경우 diffUtil로 notifyFataSetChanged 대체 가능
    class CurrentHabitComparator: DiffUtil.ItemCallback<Habit>(){
        override fun areItemsTheSame(oldItem: Habit, newItem: Habit): Boolean {
            return oldItem === newItem
        }

        override fun areContentsTheSame(oldItem: Habit, newItem: Habit): Boolean {
            return oldItem.id == newItem.id
        }
    }
}